File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Classes/Tables/RobotInfo.java
package com.alphadevelopmentsolutions.frcscout.Classes.Tables;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.alphadevelopmentsolutions.frcscout.Classes.Database;

import java.util.ArrayList;

public class RobotInfo extends Table
{

    public static final String TABLE_NAME = "robot_info";
    public static final String COLUMN_NAME_ID = "Id";
    public static final String COLUMN_NAME_YEAR_ID = "YearId";
    public static final String COLUMN_NAME_EVENT_ID = "EventId";
    public static final String COLUMN_NAME_TEAM_ID = "TeamId";

    public static final String COLUMN_NAME_PROPERTY_STATE = "PropertyState";
    public static final String COLUMN_NAME_PROPERTY_KEY = "PropertyKey";
    public static final String COLUMN_NAME_PROPERTY_VALUE = "PropertyValue";

    public static final String COLUMN_NAME_IS_DRAFT = "IsDraft";

    public static final String CREATE_TABLE =
            "CREATE TABLE " + TABLE_NAME +" (" +
                    COLUMN_NAME_ID + " INTEGER PRIMARY KEY," +
                    COLUMN_NAME_YEAR_ID + " INTEGER," +
                    COLUMN_NAME_EVENT_ID + " TEXT," +
                    COLUMN_NAME_TEAM_ID + " INTEGER," +

                    COLUMN_NAME_PROPERTY_STATE + " TEXT," +
                    COLUMN_NAME_PROPERTY_KEY + " TEXT," +
                    COLUMN_NAME_PROPERTY_VALUE + " TEXT," +

                    COLUMN_NAME_IS_DRAFT + " INTEGER)";

    private int id;
    private int yearId;
    private String eventId;
    private int teamId;

    private String propertyState;
    private String propertyKey;
    private String propertyValue;

    private boolean isDraft;

    public RobotInfo(
            int id,
            int yearId,
            String eventId,
            int teamId,

            String propertyState,
            String propertyKey,
            String propertyValue,

            boolean isDraft)
    {
        this.id = id;
        this.yearId = yearId;
        this.eventId = eventId;
        this.teamId = teamId;

        this.propertyState = propertyState;
        this.propertyKey = propertyKey;
        this.propertyValue = propertyValue;

        this.isDraft = isDraft;
    }

    /**
     * Used for loading
     * @param id to load
     */
    public RobotInfo(int id)
    {
        this.id = id;
    }

    //region Getters


    public int getId()
    {
        return id;
    }

    public int getYearId()
    {
        return yearId;
    }

    public String getEventId()
    {
        return eventId;
    }

    public int getTeamId()
    {
        return teamId;
    }

    public String getPropertyState()
    {
        return propertyState;
    }

    public String getPropertyKey()
    {
        return propertyKey;
    }

    public String getPropertyValue()
    {
        return propertyValue;
    }

    public boolean isDraft()
    {
        return isDraft;
    }

    @Override
    public String toString()
    {
        return "Team " + getTeamId() + " - Robot Info";
    }

    //endregion

    //region Setters

    public void setId(int id)
    {
        this.id = id;
    }

    public void setYearId(int yearId)
    {
        this.yearId = yearId;
    }

    public void setEventId(String eventId)
    {
        this.eventId = eventId;
    }

    public void setTeamId(int teamId)
    {
        this.teamId = teamId;
    }

    public void setPropertyState(String propertyState)
    {
        this.propertyState = propertyState;
    }

    public void setPropertyKey(String propertyKey)
    {
        this.propertyKey = propertyKey;
    }

    public void setPropertyValue(String propertyValue)
    {
        this.propertyValue = propertyValue;
    }

    public void setDraft(boolean draft)
    {
        isDraft = draft;
    }


    //endregion

    //region Load, Save & Delete

    /**
     * Loads the object from the database and populates all values
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    public boolean load(Database database)
    {
        //try to open the DB if it is not open
        if(!database.isOpen()) database.open();

        if(database.isOpen())
        {
            ArrayList<RobotInfo> robotInfoList = getRobotInfo(null, null, null, this, false, database);
            RobotInfo robotInfo = (robotInfoList.size() > 0 ) ? robotInfoList.get(0) : null;

            if (robotInfo != null)
            {
                setYearId(robotInfo.getYearId());
                setEventId(robotInfo.getEventId());
                setTeamId(robotInfo.getTeamId());

                setPropertyState(robotInfo.getPropertyState());
                setPropertyKey(robotInfo.getPropertyKey());
                setPropertyValue(robotInfo.getPropertyValue());

                setDraft(robotInfo.isDraft());
                return true;
            }
        }

        return false;
    }

    /**
     * Saves the object into the database
     * @param database used for interacting with the SQLITE db
     * @return int id of the saved object
     */
    public int save(Database database)
    {
        int id = -1;

        //try to open the DB if it is not open
        if(!database.isOpen())
            database.open();

        if(database.isOpen())
            id = (int) database.setRobotInfo(this);

        //set the id if the save was successful
        if(id > 0)
            setId(id);

        return getId();
    }

    /**
     * Deletes the object from the database
     * @param database used for interacting with the SQLITE db
     * @return boolean if successful
     */
    public boolean delete(Database database)
    {
        boolean successful = false;

        //try to open the DB if it is not open
        if(!database.isOpen()) database.open();

        if(database.isOpen())
        {
            successful = database.deleteRobotInfo(this);

        }

        return successful;
    }

    /**
     * Clears all data from the classes table
     * @param database used to clear table
     * @param clearDrafts boolean if you want to include drafts in the clear
     */
    public static void clearTable(Database database, boolean clearDrafts)
    {
        database.clearTable(TABLE_NAME, clearDrafts);
    }

    /**
     * Returns arraylist of pit cards with specified filters from database
     * @param year if specified, filters by year id
     * @param event if specified, filters by event id
     * @param team if specified, filters by team id
     * @param robotInfo if specified, filters by robotinfo id
     * @param onlyDrafts if true, filters by draft
     * @param database used to load
     * @return arraylist of robotInfo
     */
    public static ArrayList<RobotInfo> getRobotInfo(@Nullable Year year, @Nullable Event event, @Nullable Team team, @Nullable RobotInfo robotInfo, boolean onlyDrafts, @NonNull Database database)
    {
        return database.getRobotInfo(year, event, team, robotInfo, onlyDrafts);
    }

    //endregion
}

File: app/src/main/java/com/alphadevelopmentsolutions/frcscout/Classes/Tables/RobotInfo.kt
package com.alphadevelopmentsolutions.frcscout.Classes.Tables

import com.alphadevelopmentsolutions.frcscout.Classes.Database
import com.alphadevelopmentsolutions.frcscout.Classes.MasterContentValues
import com.alphadevelopmentsolutions.frcscout.Classes.TableColumn
import com.alphadevelopmentsolutions.frcscout.Interfaces.ChildTableCompanion
import com.alphadevelopmentsolutions.frcscout.Interfaces.SQLiteDataTypes
import java.util.*

class RobotInfo(
        localId: Long = DEFAULT_LONG,
        serverId: Long = DEFAULT_LONG,
        var yearId: Long = DEFAULT_LONG,
        var eventId: String = DEFAULT_STRING,
        var teamId: Long = DEFAULT_LONG,
        var propertyValue: String= DEFAULT_STRING,
        var propertyKeyId: Long = DEFAULT_LONG,
        var isDraft: Boolean = DEFAULT_BOOLEAN) : Table(TABLE_NAME, localId, serverId)
{
    companion object: ChildTableCompanion
    {
        override val TABLE_NAME = "robot_info"

        const val COLUMN_NAME_YEAR_ID = "YearId"
        const val COLUMN_NAME_EVENT_ID = "EventId"
        const val COLUMN_NAME_TEAM_ID = "TeamId"
        const val COLUMN_NAME_PROPERTY_VALUE = "PropertyValue"
        const val COLUMN_NAME_PROPERTY_KEY_ID = "PropertyKeyId"

        override val childColumns: ArrayList<TableColumn>
            get() = ArrayList<TableColumn>().apply {
                add(TableColumn(COLUMN_NAME_YEAR_ID, SQLiteDataTypes.INTEGER))
                add(TableColumn(COLUMN_NAME_EVENT_ID, SQLiteDataTypes.TEXT))
                add(TableColumn(COLUMN_NAME_TEAM_ID, SQLiteDataTypes.INTEGER))
                add(TableColumn(COLUMN_NAME_PROPERTY_VALUE, SQLiteDataTypes.TEXT))
                add(TableColumn(COLUMN_NAME_PROPERTY_KEY_ID, SQLiteDataTypes.INTEGER))
                add(TableColumn(COLUMN_NAME_IS_DRAFT, SQLiteDataTypes.INTEGER))
            }
        
        /**
         * Returns [ArrayList] of [RobotInfo] with specified filters from [database]
         * @param year if specified, filters by [year] id
         * @param event if specified, filters by [event] id
         * @param team if specified, filters by [team] id
         * @param robotInfo if specified, filters by [robotInfo] id
         * @param onlyDrafts if true, filters by [isDraft]
         * @param database used to load
         * @return [ArrayList] of [RobotInfo]
         */
        fun getObjects(year: Year?, event: Event?, team: Team?, robotInfoKey: RobotInfoKey?, robotInfo: RobotInfo?, onlyDrafts: Boolean, database: Database): ArrayList<RobotInfo> {
            return ArrayList<RobotInfo>().apply {

                val whereStatement = StringBuilder()
                val whereArgs = ArrayList<String>()

                //filter by object
                if (year != null)
                {
                    whereStatement.append("$COLUMN_NAME_YEAR_ID = ? ")
                    whereArgs.add(year.serverId.toString())
                }

                //filter by object
                if (event != null)
                {
                    whereStatement.append(if (whereStatement.isNotEmpty()) " AND " else "").append("$COLUMN_NAME_EVENT_ID = ? ")
                    whereArgs.add(event.blueAllianceId)
                }

                //filter by object
                if (team != null)
                {
                    whereStatement.append(if (whereStatement.isNotEmpty()) " AND " else "").append("$COLUMN_NAME_TEAM_ID = ? ")
                    whereArgs.add(team.serverId.toString())
                }

                //filter by object
                if (robotInfoKey != null)
                {
                    whereStatement.append(if (whereStatement.isNotEmpty()) " AND " else "").append("$COLUMN_NAME_PROPERTY_KEY_ID = ? ")
                    whereArgs.add(robotInfoKey.serverId.toString())
                }

                //filter by object
                if (robotInfo != null)
                {
                    whereStatement.append(if (whereStatement.isNotEmpty()) " AND " else "").append("$COLUMN_NAME_LOCAL_ID = ? ")
                    whereArgs.add(robotInfo.localId.toString())
                }

                //filter by object
                if (onlyDrafts)
                {
                    whereStatement.append(if (whereStatement.isNotEmpty()) " AND " else "").append("$COLUMN_NAME_IS_DRAFT = 1 ")
                }

                //add all object records to array list
                with(database.getObjects(
                        TABLE_NAME,
                        whereStatement.toString(),
                        whereArgs))
                {
                    if (this != null) {
                        while (moveToNext()) {
                            add(
                                    RobotInfo(
                                            getLong(COLUMN_NAME_LOCAL_ID),
                                            getLong(COLUMN_NAME_SERVER_ID),
                                            getLong(COLUMN_NAME_YEAR_ID),
                                            getString(COLUMN_NAME_EVENT_ID),
                                            getLong(COLUMN_NAME_TEAM_ID),
                                            getString(COLUMN_NAME_PROPERTY_VALUE),
                                            getLong(COLUMN_NAME_PROPERTY_KEY_ID),
                                            getBoolean(COLUMN_NAME_IS_DRAFT)
                                    )
                            )
                        }

                        close()
                    }
                }
            }
        }
    }

    /**
     * @see Table.load
     */
    override fun load(database: Database): Boolean
    {
        with(getObjects(null, null, null, null, this, false, database))
        {
            with(if (size > 0) this[0] else null)
            {
                if (this != null)
                {
                    loadParentValues(this)
                    this@RobotInfo.yearId = yearId
                    this@RobotInfo.eventId = eventId
                    this@RobotInfo.teamId = teamId
                    this@RobotInfo.propertyValue = propertyValue
                    this@RobotInfo.propertyKeyId = propertyKeyId
                    this@RobotInfo.isDraft = isDraft
                    return true
                }
            }
        }

        return false
    }

    /**
     * @see Table.childValues
     */
    override val childValues: MasterContentValues
        get() = MasterContentValues().apply {
            put(COLUMN_NAME_YEAR_ID, yearId)
            put(COLUMN_NAME_EVENT_ID, eventId)
            put(COLUMN_NAME_TEAM_ID, teamId)
            put(COLUMN_NAME_PROPERTY_VALUE, propertyValue)
            put(COLUMN_NAME_PROPERTY_KEY_ID, propertyKeyId)
            put(COLUMN_NAME_IS_DRAFT, isDraft)
        }

    /**
     * @see Table.toString
     */
    override fun toString(): String
    {
        return "Team $teamId - Robot Info"
    }
}
